### ISSUE TITLE
twisted.application.internet generates an extra entry in its __all__ list.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4373]
|<img alt="kench's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kench reported|
|-|-|
|Trac ID|trac#4373|
|Type|defect|
|Created|2010-03-18 21:44:56Z|
|Branch|https://github.com/twisted/twisted/tree/internet-all-list-4373|

twisted.application.internet dynamically generates xxxServer and xxxClient classes for each of a list of supported transports and then generates its ```__all__``` list similarly.  The difference is that it bypasses generating the class for MulticastClient but does not bypass adding an entry to ```__all__``` for the non-existent class.

This can be demonstrated by executing the following:

```python
>>> from twisted.application.internet import *
Traceback (most recent call last):
  File "<string>", line 1, in <fragment>
AttributeError: 'module' object has no attribute 'MulticastClient'
>>>
```
 

Attachments:

* [twisted_4373_test.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/72f/72f8ca1a41999fa6f478762ed2820147c7c236eb/80a6b86886237cea27d2f4253a51690eb5282959.diff) (810 bytes) - added by ltaylor.volks on 2010-03-22 20:15:52Z - Test to ensure subclasses exposed by __all__ are valid attributes of twisted.application.internet
* [twisted-4373.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/72f/72f8ca1a41999fa6f478762ed2820147c7c236eb/0a0db9c7ac410424092ad837b4ffa09f23ab71a1.diff) (1371 bytes) - added by ltaylor.volks on 2010-03-26 00:19:04Z - Patch addresses the ticket using solution #2. Test ensures subclasses exposed by !__all!__ are valid attributes of twisted.application.internet. (Ignore twisted_4373_test.diff)

<details><summary>Searchable metadata</summary>

```
trac-id__4373 4373
type__defect defect
reporter__kench kench
priority__low low
milestone__ 
branch__branches_internet_all_list_4373 branches/internet-all-list-4373
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1268948696000000 1268948696000000
changetime__1269882833000000 1269882833000000
version__None None
owner__ 
cc__ltaylor.volks cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kench's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|kench commented|
|-|-|

Possible solutions:

(1) ```__all__```.remove(_MulticastClient_) after the list is generated.

-- or --

(2) Remove ```' Multicast'``` from the end of the string in the ```'for tran in ...'``` line so it doesn't generate the Server/Client pair for Multicast, and add ```'MulticastServer'``` to the list individually.

-- or --

(3) Filter ```'MulticastClient'``` out of the list comprehension by adding ```if tran+side != 'MulticastClient'``` to it.

-- or --

(4) Build the ```__all__``` list as the classes themselves are generated.

-- or --

(5) Define a dummy class for MulticastClient that raises NotImplementedError and/or insults the user's intelligence when he tries to use it.  (That is what the insults package is for, right? ... uh, nevermind.)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra removed owner|
|-|-|

Thanks for the patch. Putting it up for review as described on ReviewProcess.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess commented|
|-|-|

(In [28783]) Branching to 'internet-all-list-4373'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess commented|
|-|-|

(In [28784]) Apply twisted-4373.diff

Also some cosmetic changes: use assertTrue instead of assert_ in test_importAll and
test_everythingThere (ltaylor.volks, there's more in this in the [test standard](http://twistedmatrix.com/documents/current/core/development/policy/test-standard.html)), remove unused win32 skip check, remove trailing whitespace, add epytext markup to test_importAll and docstring to test_everythingThere.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

Thanks for working on this, ltaylor.volks! The [builds](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/internet-all-list-4373) look good and I think it's ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess set status to `closed`|
|-|-|

(In [28785]) Authors: ltaylor.volks, jesstess
Reviewer: jesstess
Fixes: [#4373](https://github.com/twisted/twisted/issues/4373)

twisted.application.internet no longer generates an extra and invalid
entry in its __all__ list for the nonexistent MulticastClient.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
